File: go/pondsocket/endpoint.go
// This file contains the Endpoint struct which manages WebSocket connections for a specific path.
// Endpoints handle connection lifecycle, message routing between clients and channels,
// and coordinate join/leave operations for channel-based communication.
package pondsocket

import (
	"context"
	"errors"
	"sync"

	"github.com/google/uuid"
)

type Endpoint struct {
	path        string
	connections *store[Transport]
	middleware  *middleware[joinEvent, interface{}]
	channels    *store[*Channel]
	options     *Options
	ctx         context.Context
	mutex       sync.RWMutex
}

type joinEvent struct {
	user      Transport
	channel   string
	requestId string
	payload   interface{}
}

func newEndpoint(ctx context.Context, path string, options *Options) *Endpoint {
	return &Endpoint{
		path:        path,
		options:     options,
		connections: newStore[Transport](),
		channels:    newStore[*Channel](),
		middleware:  newMiddleWare[joinEvent, interface{}](),
		ctx:         ctx,
	}
}

// CreateChannel creates a new channel pattern for this endpoint.
// The path parameter defines the channel pattern (e.g., "room/:id" or "user/:userId/notifications").
// The handlerFunc is called when a client attempts to join a matching channel,
// allowing for custom authorization and setup logic.
// Returns a Lobby instance for further configuration of message and presence handling.
func (e *Endpoint) CreateChannel(path string, handlerFunc JoinEventHandler) *Lobby {
	if err := e.checkState(); err != nil {
		return newLobby(e)
	}
	lobby := newLobby(e)

	e.middleware.Use(func(ctx context.Context, request joinEvent, _ interface{}, next nextFunc) error {
		if err := e.checkState(); err != nil {
			return err
		}
		select {
		case <-ctx.Done():
			return ctx.Err()

		default:
		}
		route, err := parse(path, request.channel)

		if err != nil {
			return next()
		}
		ch, err := lobby.getOrCreateChannel(request.channel)

		if err != nil {
			return wrapF(err, "failed to get or create channel instance for %s", request.channel)
		}
		ctx, cancel := context.WithTimeout(e.ctx, e.options.InternalQueueTimeout)

		defer cancel()

		ev := &Event{
			Action:      system,
			ChannelName: request.channel,
			RequestId:   request.requestId,
			Event:       string(joinChannelEvent),
			Payload:     request.payload,
		}
		joinCtx := newJoinContext(ctx, ch, route, request.user, ev)

		if err := handlerFunc(joinCtx); err != nil {
			return err
		}
		if joinCtx.err != nil {
			return joinCtx
		}
		return nil
	})

	return lobby
}

// CloseConnection forcefully closes one or more connections by their IDs.
// This can be used to disconnect specific clients from the server.
// Returns an error if any of the specified connections cannot be found or closed.
// If no IDs are provided, the method returns nil without performing any action.
func (e *Endpoint) CloseConnection(ids ...string) error {
	if err := e.checkState(); err != nil {
		return err
	}
	if len(ids) == 0 {
		return nil
	}
	var errs []error
	for _, id := range ids {
		if _, err := e.connections.Read(id); err != nil {
			errs = append(errs, wrapF(err, "connection %s not found", id))
		}
	}

	connsToClose := e.connections.GetByKeys(ids...)

	closeErr := mapToError(connsToClose, func(conn Transport) error {
		conn.Close()

		return e.connections.Delete(conn.GetID())
	})

	errs = append(errs, closeErr)

	return combine(errs...)
}

// GetClients returns a slice of active connections for this endpoint.
// If no IDs are provided, it returns all active connections.
// If specific IDs are provided, it returns only the connections matching those IDs.
// Non-existent IDs are silently ignored.
func (e *Endpoint) GetClients(ids ...string) []Transport {
	if err := e.checkState(); err != nil {
		return []Transport{}
	}
	if len(ids) == 0 {
		clients := e.connections.Values()

		return clients.items
	}
	clients := e.connections.GetByKeys(ids...)

	return clients.items
}

// GetChannelByName retrieves a channel instance by its exact name.
// Returns the channel if it exists, or an error if the channel is not found.
// This method looks up active channels that have been created through client joins.
func (e *Endpoint) GetChannelByName(name string) (*Channel, error) {
	if err := e.checkState(); err != nil {
		return nil, err
	}
	channel, err := e.channels.Read(name)

	return channel, err
}

func (e *Endpoint) joinChannel(ev *Event, user Transport) error {
	if err := e.checkState(); err != nil {
		return err
	}
	ctx, cancel := context.WithTimeout(e.ctx, e.options.InternalQueueTimeout)

	defer cancel()

	joinReq := joinEvent{
		user:      user,
		requestId: ev.RequestId,
		channel:   ev.ChannelName,
		payload:   ev.Payload,
	}
	err := e.middleware.Handle(ctx, joinReq, nil, e.notFoundHandler(ctx, ev.RequestId))

	if err != nil {
		errMsg := Event{
			Action:      system,
			ChannelName: ev.ChannelName,
			RequestId:   ev.RequestId,
			Event:       string(internalErrorEvent),
			Payload:     internal(ev.ChannelName, "Failed to process join request"),
		}

		var pondErr *Error
		if errors.As(err, &pondErr) {
			errMsg.Payload = pondErr
		}
		_ = user.SendJSON(errMsg)
	}
	return err
}

func (e *Endpoint) leaveChannel(ev *Event, user Transport) error {
	if err := e.checkState(); err != nil {
		return err
	}
	currentChannel, err := e.channels.Read(ev.ChannelName)

	if err != nil {
		_ = user.SendJSON(Event{
			Action:      system,
			ChannelName: ev.ChannelName,
			RequestId:   ev.RequestId,
			Event:       string(notFoundEvent),
			Payload:     notFound(ev.ChannelName, "Channel not found"),
		})

		return err
	}
	if err = currentChannel.RemoveUser(user.GetID(), "explicit_leave"); err != nil {
		_ = user.SendJSON(Event{
			Action:      system,
			ChannelName: ev.ChannelName,
			RequestId:   ev.RequestId,
			Event:       string(internalErrorEvent),
			Payload:     wrapF(err, "failed to leave channel"),
		})

		return err
	}
	_ = user.SendJSON(Event{
		Action:      system,
		ChannelName: ev.ChannelName,
		RequestId:   ev.RequestId,
		Event:       string(exitAcknowledgeEvent),
		Payload:     make(map[string]interface{}),
	})

	return nil
}

func (e *Endpoint) broadcastMessage(ev *Event, user Transport) error {
	if err := e.checkState(); err != nil {
		return err
	}
	currentChannel, err := e.channels.Read(ev.ChannelName)

	if err != nil {
		_ = user.SendJSON(Event{
			Action:      system,
			ChannelName: ev.ChannelName,
			RequestId:   ev.RequestId,
			Event:       string(notFoundEvent),
			Payload:     notFound(ev.ChannelName, "Channel not found"),
		})

		return err
	}
	if err = currentChannel.broadcast(user.GetID(), ev); err != nil {
		_ = user.SendJSON(Event{
			Action:      system,
			ChannelName: ev.ChannelName,
			RequestId:   ev.RequestId,
			Event:       string(internalErrorEvent),
			Payload:     wrapF(err, "failed to process broadcast"),
		})

		return err
	}
	return nil
}

func (e *Endpoint) notFoundHandler(ctx context.Context, requestID string) FinalHandlerFunc[joinEvent, interface{}] {
	return func(request joinEvent, response interface{}) error {
		select {
		case <-ctx.Done():
			return ctx.Err()

		default:
		}
		if err := e.checkState(); err != nil {
			return err
		}
		message := Event{
			Action:      system,
			ChannelName: request.channel,
			RequestId:   requestID,
			Event:       string(notFoundEvent),
			Payload:     notFound(request.channel, "No channel handler found for this topic"),
		}
		return request.user.SendJSON(message)
	}
}

func (e *Endpoint) handleMessage() transportEventHandler {
	return func(ev Event, user Transport) error {
		if err := e.checkState(); err != nil {
			return err
		}
		switch ev.Action {
		case joinChannelEvent:
			return e.joinChannel(&ev, user)

		case leaveChannelEvent:
			return e.leaveChannel(&ev, user)

		case broadcast:
			return e.broadcastMessage(&ev, user)

		default:
			_ = user.SendJSON(Event{
				Action:      system,
				ChannelName: ev.ChannelName,
				RequestId:   ev.RequestId,
				Event:       string(notFoundEvent),
				Payload:     notFound(ev.ChannelName, "Unknown or unsupported event type").withDetails(map[string]string{"eventType": ev.Event}),
			})

			return badRequest(ev.ChannelName, "Unknown event type").withDetails(map[string]string{"eventType": ev.Event})
		}
	}
}

func (e *Endpoint) handleClose(c Transport) error {
	if e.options.Hooks != nil && e.options.Hooks.OnDisconnect != nil {
		e.options.Hooks.OnDisconnect(c)
	}
	if e.options.Hooks != nil && e.options.Hooks.Metrics != nil {
		e.options.Hooks.Metrics.ConnectionClosed(c.GetID(), 0)
	}
	return e.connections.Delete(c.GetID())
}

func (e *Endpoint) addConnection(conn Transport) error {
	if err := e.checkState(); err != nil {
		conn.Close()

		return err
	}
	if e.options.Hooks != nil && e.options.Hooks.OnConnect != nil {
		if err := e.options.Hooks.OnConnect(conn); err != nil {
			conn.Close()

			return wrapF(err, "OnConnect hook failed")
		}
	}
	if e.options.MaxConnections > 0 {
		e.mutex.RLock()

		currentCount := e.connections.Len()

		e.mutex.RUnlock()

		if currentCount >= e.options.MaxConnections {
			conn.Close()

			return unavailable(string(gatewayEntity), "Maximum connections reached")
		}
	}
	if err := e.connections.Create(conn.GetID(), conn); err != nil {
		conn.Close()

		return wrapF(err, "failed to store connection %s", conn.GetID())
	}
	if e.options.Hooks != nil && e.options.Hooks.Metrics != nil {
		e.options.Hooks.Metrics.ConnectionOpened(conn.GetID(), "endpoint")
	}
	conn.OnMessage(e.handleMessage())

	conn.OnClose(e.handleClose)

	conn.HandleMessages()

	err := conn.SendJSON(Event{
		Action:      connect,
		ChannelName: string(gatewayEntity),
		RequestId:   uuid.NewString(),
		Event:       string(connectionEvent),
		Payload: map[string]interface{}{
			"connectionId": conn.GetID(),
		},
	})

	if err != nil {
		conn.Close()

		_ = e.connections.Delete(conn.GetID())

		if e.options.Hooks != nil && e.options.Hooks.Metrics != nil {
			e.options.Hooks.Metrics.ConnectionError(conn.GetID(), err)
		}
		return wrapF(err, "failed to send connection confirmation to %s", conn.GetID())
	}
	return nil
}

func (e *Endpoint) sendMessage(userId string, event Event) error {
	userConn, err := e.connections.Read(userId)

	if err != nil {
		return notFound(string(gatewayEntity), "User not found").withDetails(map[string]string{"userId": userId})
	}
	return userConn.SendJSON(event)
}

func (e *Endpoint) pushMessage(connID string, data []byte) error {
	if err := e.checkState(); err != nil {
		return err
	}

	conn, err := e.connections.Read(connID)
	if err != nil {
		return notFound(string(gatewayEntity), "Connection not found").withDetails(map[string]string{"connId": connID})
	}

	if conn.Type() != TransportSSE {
		return badRequest(string(gatewayEntity), "PushMessage only supported for SSE connections")
	}

	return conn.PushMessage(data)
}

func (e *Endpoint) checkState() error {
	select {
	case <-e.ctx.Done():
		err := e.ctx.Err()

		return wrapF(err, "endpoint is shutting down")

	default:
		return nil
	}
}

File: rust/pondsocket/src/endpoint.rs
use async_trait::async_trait;
use pondsocket_common::uuid;
use serde_json::json;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Instant;
use tokio::sync::RwLock;

use crate::contexts::{ConnectionContext, ConnectionDecision, IncomingConnection, JoinContext};
use crate::errors::{Result, conflict, forbidden, not_found};
use crate::lobby::Lobby;
use crate::parser::parse;
use crate::pubsub::PubSub;
use crate::transport::Transport;
use crate::types::{Event, Options, Route, User};

#[async_trait]
pub trait ConnectionHandler: Send + Sync {
    async fn call(&self, ctx: &mut ConnectionContext) -> Result<()>;
}

#[async_trait]
impl<F, Fut> ConnectionHandler for F
where
    F: Fn(&mut ConnectionContext) -> Fut + Send + Sync,
    Fut: std::future::Future<Output = Result<()>> + Send,
{
    async fn call(&self, ctx: &mut ConnectionContext) -> Result<()> {
        self(ctx).await
    }
}

#[async_trait]
pub trait JoinHandler: Send + Sync {
    async fn call(&self, ctx: &mut JoinContext) -> Result<()>;
}

#[async_trait]
impl<F, Fut> JoinHandler for F
where
    F: Fn(&mut JoinContext) -> Fut + Send + Sync,
    Fut: std::future::Future<Output = Result<()>> + Send,
{
    async fn call(&self, ctx: &mut JoinContext) -> Result<()> {
        self(ctx).await
    }
}

struct ChannelRegistration {
    pattern: String,
    handler: Arc<dyn JoinHandler>,
    lobby: Arc<Lobby>,
}

pub struct Endpoint {
    path: String,
    connection_handler: Arc<dyn ConnectionHandler>,
    options: Options,
    pubsub: Option<Arc<dyn PubSub>>,
    connections: RwLock<HashMap<String, Arc<dyn Transport>>>,
    channel_registrations: RwLock<Vec<ChannelRegistration>>,
    connect_times: RwLock<HashMap<String, Instant>>,
}

impl Endpoint {
    pub fn new<H>(
        path: impl Into<String>,
        connection_handler: H,
        options: Options,
        pubsub: Option<Arc<dyn PubSub>>,
    ) -> Arc<Self>
    where
        H: ConnectionHandler + 'static,
    {
        Arc::new(Self {
            path: path.into(),
            connection_handler: Arc::new(connection_handler),
            options,
            pubsub,
            connections: RwLock::new(HashMap::new()),
            channel_registrations: RwLock::new(Vec::new()),
            connect_times: RwLock::new(HashMap::new()),
        })
    }

    pub fn path(&self) -> &str {
        &self.path
    }

    pub async fn create_channel<H>(
        self: &Arc<Self>,
        pattern: impl Into<String>,
        handler: H,
    ) -> Arc<Lobby>
    where
        H: JoinHandler + 'static,
    {
        let pattern = pattern.into();
        let lobby = Lobby::new(
            pattern.clone(),
            self.path.clone(),
            self.options.clone(),
            self.pubsub.clone(),
        );
        self.channel_registrations
            .write()
            .await
            .push(ChannelRegistration {
                pattern,
                handler: Arc::new(handler),
                lobby: lobby.clone(),
            });
        lobby
    }

    pub async fn request_connection(
        &self,
        mut incoming: IncomingConnection,
        route: Route,
        user_id: Option<String>,
    ) -> ConnectionContext {
        let uid = user_id.unwrap_or_else(uuid);
        incoming.id = uid.clone();
        let mut ctx = ConnectionContext::new(uid, incoming, route);
        if let Err(err) = self.connection_handler.call(&mut ctx).await
            && ctx.decision() == ConnectionDecision::Pending
        {
            let _ = ctx.decline(err.code, err.message);
        }
        if ctx.decision() == ConnectionDecision::Pending {
            let _ = ctx.decline(401, "Connection handler did not accept or decline");
        }
        ctx
    }

    pub async fn register_transport(&self, transport: Arc<dyn Transport>) -> Result<()> {
        let user_id = transport.id().to_owned();
        {
            let mut connections = self.connections.write().await;
            if connections.contains_key(&user_id) {
                return Err(conflict(&self.path, "connection id already registered"));
            }
            if self.options.max_connections > 0 && connections.len() >= self.options.max_connections
            {
                return Err(forbidden("GATEWAY", "Maximum connections reached"));
            }
            connections.insert(user_id.clone(), transport.clone());
        }
        self.connect_times
            .write()
            .await
            .insert(user_id.clone(), Instant::now());
        self.send_connect_event(transport.clone()).await?;
        if let Some(hooks) = &self.options.hooks {
            if let Some(metrics) = &hooks.metrics {
                metrics.connection_opened(&user_id, &self.path);
            }
            if let Some(hook) = &hooks.on_connect {
                hook.call(&user_id).await;
            }
        }
        Ok(())
    }

    pub async fn unregister_transport(&self, user_id: &str) -> Result<()> {
        let lobbies = self
            .channel_registrations
            .read()
            .await
            .iter()
            .map(|reg| reg.lobby.clone())
            .collect::<Vec<_>>();
        for lobby in lobbies {
            for channel in lobby.list_channels().await {
                if channel.has_user(user_id).await {
                    channel.remove_user(user_id, "connection_closed").await?;
                }
            }
        }
        self.connections.write().await.remove(user_id);
        let duration = self
            .connect_times
            .write()
            .await
            .remove(user_id)
            .map(|t| t.elapsed());
        if let Some(hooks) = &self.options.hooks {
            if let (Some(metrics), Some(duration)) = (&hooks.metrics, duration) {
                metrics.connection_closed(user_id, duration);
            }
            if let Some(hook) = &hooks.on_disconnect {
                hook.call(user_id).await;
            }
        }
        Ok(())
    }

    pub async fn handle_message(
        self: &Arc<Self>,
        event: Event,
        transport: Arc<dyn Transport>,
    ) -> Result<()> {
        match event.action.as_str() {
            "JOIN_CHANNEL" => self.handle_join(event, transport).await,
            "LEAVE_CHANNEL" => self.handle_leave(event, transport).await,
            "BROADCAST" => self.handle_broadcast(event, transport).await,
            _ => Ok(()),
        }
    }

    async fn handle_join(
        self: &Arc<Self>,
        event: Event,
        transport: Arc<dyn Transport>,
    ) -> Result<()> {
        for reg in self.channel_registrations.read().await.iter() {
            let Ok(route) = parse(&reg.pattern, &event.channel_name) else {
                continue;
            };
            let channel = reg.lobby.acquire_for_join(&event.channel_name).await?;
            if let Some(hooks) = &self.options.hooks
                && let Some(before) = &hooks.before_join
            {
                let user = User {
                    id: transport.id().to_owned(),
                    assigns: transport.clone_assigns().await,
                    presence: None,
                };
                before.call(&user, &event.channel_name).await;
            }
            let mut ctx =
                JoinContext::new(channel.clone(), event.clone(), transport.clone(), route);
            if let Err(err) = reg.handler.call(&mut ctx).await
                && !ctx.has_responded()
            {
                let _ = ctx.decline(err.code, err.message).await;
            }
            let mut outcome = Ok(());
            if !ctx.has_responded() {
                outcome = ctx.decline(401, "Join handler did not respond").await;
            }
            if ctx.is_accepted()
                && let Some(hooks) = &self.options.hooks
                && let Some(after) = &hooks.after_join
            {
                let user = User {
                    id: transport.id().to_owned(),
                    assigns: transport.clone_assigns().await,
                    presence: None,
                };
                after.call(&user, &event.channel_name).await;
            }
            reg.lobby.finish_join(channel).await;
            return outcome;
        }
        let channel_name = event.channel_name.clone();
        transport
            .send_event(Event::new(
                "SYSTEM",
                channel_name.clone(),
                event.request_id,
                "NOT_FOUND",
                json!({ "channel": channel_name }),
            ))
            .await
    }

    async fn handle_leave(&self, event: Event, transport: Arc<dyn Transport>) -> Result<()> {
        if let Some(channel) = self.find_channel(&event.channel_name).await {
            channel
                .remove_user(transport.id(), "explicit_leave")
                .await?;
            transport
                .send_event(Event::new(
                    "SYSTEM",
                    event.channel_name,
                    event.request_id,
                    "EXIT_ACKNOWLEDGE",
                    json!({}),
                ))
                .await?;
        }
        Ok(())
    }

    async fn handle_broadcast(
        self: &Arc<Self>,
        event: Event,
        transport: Arc<dyn Transport>,
    ) -> Result<()> {
        let Some(channel) = self.find_channel(&event.channel_name).await else {
            let channel_name = event.channel_name.clone();
            return transport
                .send_event(Event::new(
                    "SYSTEM",
                    channel_name.clone(),
                    event.request_id,
                    "NOT_FOUND",
                    json!({ "channel": channel_name }),
                ))
                .await;
        };
        channel.handle_incoming_event(event, transport.id()).await
    }

    async fn find_channel(&self, name: &str) -> Option<Arc<crate::channel::Channel>> {
        for reg in self.channel_registrations.read().await.iter() {
            if let Some(channel) = reg.lobby.get_channel(name).await {
                return Some(channel);
            }
        }
        None
    }

    async fn send_connect_event(&self, transport: Arc<dyn Transport>) -> Result<()> {
        let id = transport.id().to_owned();
        transport
            .send_event(Event::new(
                "CONNECT",
                "GATEWAY",
                uuid(),
                "CONNECTION",
                json!({ "id": id, "connectionId": id }),
            ))
            .await
    }

    pub async fn get_transport(&self, user_id: &str) -> Option<Arc<dyn Transport>> {
        self.connections.read().await.get(user_id).cloned()
    }

    pub async fn close_connection(&self, user_id: &str) -> Result<()> {
        let Some(transport) = self.get_transport(user_id).await else {
            return Err(not_found(&self.path, "connection not found"));
        };
        transport.close().await
    }

    pub async fn close(&self) -> Result<()> {
        let lobbies = self
            .channel_registrations
            .read()
            .await
            .iter()
            .map(|reg| reg.lobby.clone())
            .collect::<Vec<_>>();
        for lobby in lobbies {
            lobby.close().await?;
        }
        let transports = self
            .connections
            .read()
            .await
            .values()
            .cloned()
            .collect::<Vec<_>>();
        for transport in transports {
            let _ = transport.close().await;
        }
        self.connections.write().await.clear();
        self.connect_times.write().await.clear();
        Ok(())
    }

    pub fn max_message_size(&self) -> usize {
        self.options.max_message_size
    }
}
